### ISSUE TITLE
Mixing out attachment timelines loses the deform

### ISSUE THREAD

**Issue by NathanSweet** [EsotericSoftware/spine-editor#445]
*Scenario:* Setting animations `a->b` where both animations have 1) attachment keys showing the same mesh, which is hidden in setup mode, and 2) deform keys.

*Result:* When `a` is applied we are mixing the entry out and so the attachment is cleared, which also clears the slot's deform. When `b` is applied, there is no deform so it mixes from the setup pose (ie no deform) to `b`.

[Forum post](http://esotericsoftware.com/forum/Mix-animation-include-setup-pose-11602)

The code in `AttachmentTimeline` that clears the attachment (and deform) when mixing out: [Animation:918](https://github.com/EsotericSoftware/spine-runtimes/blob/3.7/spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/Animation.java#L918).

When applying the deform timeline for `b`, if the slot has no current deform, the setup pose is used (equivalent to using a deform of all zeroes): [Animation:1005](https://github.com/EsotericSoftware/spine-runtimes/blob/3.7/spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/Animation.java#L1005).

In the scenario described above, `attachmentThreshold=0` (the default). If we were keeping attachments that are being mixed out (`attachmentThreshold=1`) then this line would use MixDirection `in`: [AnimationState:316](https://github.com/EsotericSoftware/spine-runtimes/blob/3.7/spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/AnimationState.java#L316). This means `AttachmentTimeline` would not clear the attachment when `a` is applied, the deform for `a` is applied, then the deform for `b` is mixed correctly.

TLDR; the problem is that attachment timelines being mixed out clear the attachment, which clears the deform.

**Comment by NathanSweet** [comment]
## Possible solutions

### AnimationState.Entry#timelineMode
If we knew when applying an attachment timeline that a later applied attachment timeline would set the attachment for the same slot, then we could avoid mixing out `a`. We have `timelineMode` (for dipping) but this isn't really the right information. We could collect more information in `setTimelineModes` and store a bit flag in the `timelineMode` int. However, it is unfortunate to need to compute/store this information.

### Slot#clearAttachment
The deform is stored on the slot because attachments are stateless and may be shared. Slot could have `clearAttachment(boolean keepDeform)`. The slot would remember what attachment the deform is for and use that deform if the same attachment is set later. `DeformTimeline` would need to set its deform based on the slot's deform attachment, not the slot's current attachment (which may be null).

### Defer slot attachment changes
`AttachmentTimeline` calls Skeleton `getAttachment` with the attachment name, resulting in 1 or 2 map lookups which have been shown to be a hotspot in the past. These lookups are wasted when subsequent timelines set the attachment for the same slot. It's been discussed before to allow the slot to take an attachment name which only gets resolved later (after animations are applied or possibly as late as render time), ensuring the minimal number of lookups.

Deferring slot attachment changes in this way could possibly help us avoid clearing the deform. We still have the problem in the scenario above where `a` is mixing out, and so its attachment timeline clears the slot attachment (or slot attachment name). Then the deform timeline for `a` does not set the deform, because the slot does not have the right vertex attachment. Somehow the deform timeline needs to know that its deform is still needed, and the slot has to know what attachment the deform is for so when that attachment is set later, the deform is used (or discarded for other attachments).

### Separate slot current attachment and deform
A slot could have a current attachment and deform, as it does now, but also a "deform attachment" which is the attachment that the deform is for. When setting the current attachment to non-null, it also sets the deform attachment and clears the deform if it doesn't match. When setting the current attachment to null, the deform and deform attachment are not affected.

This would allow a slot to have no current attachment, but to still have a deform and deform attachment. In the scenario above, when applying `a`, even though it is mixing out, the attachment timeline would first be applied as normal, setting the attachment and deform attachment. Then, because we are mixing out, it would clear the current attachment, but not the deform or deform attachment. Next the `DeformTimline` for `a` would apply the deform based on the deform attachment, not current attachment. Next `b` is applied, the attachment timeline sets the slot attachment, it matches the deform attachment so the deform is kept. Next the deform timeline is applied, correctly mixing with the existing deform (from `a`).
